### PR TITLE
added test goodfeaturetotrack.cpp

### PR DESCRIPTION
**Pull request by AlexanderUsentsov**
Added missed accuracy test for goodFeaturesToTrack functions from opencv_imgproc module

### CODE REVIEW

**Comment by vpisarev** [comment]
@AlexanderUsentsov, thank you for the PR. There are still failing tests: http://pullrequest.opencv.org

**Comment by AlexanderUsentsov** [comment]
I am working on these issues.

Thank you.
Alexander

**Comment by ilya-lavrenov** [comment]
@AlexanderUsentsov, any progress?

**Comment by AlexanderUsentsov** [comment]
Hello, Ilya!

I have fixed all bugs in accuracy test from opencv_test_imgproc but can’t  fix issues in CV_FeatureDetectorTest::regressionTest() from opencv_test_features2d .
I see that result is compared with some test data from ..\opencv_extra-master\testdata/cv/features2d/feature_detectors/detector-harris.xml.gz files.  As I changed code of this function  I receive different results to compare with old code. To fix this issue I need to change detector-harris.xml.gz file.
Could you help me to generate new test data file?

Thank you,
Alexander

**Comment by vpisarev** [comment]
@AlexanderUsentsov, please, use this guide to update the performance data:
http://code.opencv.org/projects/opencv/wiki/HowToUsePerfTests#How-to-update-perf-data

the updated data should be contributed to opencv_extra from the branch with the same name ("good_feature" in this case)

**Comment by AlexanderUsentsov** [comment]
Thank you.

I updated the performance data, contributed to opencv_extra from the "good_feature" branch and opened new pull request from this branch. Unfortunately I don’t see start of building and testing . Where am I wrong?

BR,
Alexander

**Comment by mshabunin** [comment]
@AlexanderUsentsov, I've restarted the builds.

**Comment by AlexanderUsentsov** [comment]
I’ve updated test data. Will you restart the build or I need to do something for that?

BR,
Alexander

**Comment by mshabunin** [comment]
>opened new pull request from this branch

@AlexanderUsentsov, I don't see any PRs to _opencv_extra_ from you. Can you take a look?
Although your branch is used during the build, you still need to create PR in case to merge it.

**Comment by AlexanderUsentsov** [comment]
Hi!
I’ve analyzed result of testing and see that  Features2d_Detector_Harris.regression test fails because ../opencv_extra-master/testdata/cv/features2d/feature_detectors/detector-harris.xml.gz files has data for previous variant of goodFeaturesToTrack. Could you help me to generate new test data file for this test?
BR,
Alexander

**Comment by AlexanderUsentsov** [comment]
Hi!

This is a friendly reminder.

BR,
Alexander

From: Usentsov, Alexander
Sent: Tuesday, April 7, 2015 7:21 PM
To: Itseez/opencv; Itseez/opencv
Subject: RE: [opencv] added test goodfeaturetotrack.cpp (#3540)

Hi!
I’ve analyzed result of testing and see that  Features2d_Detector_Harris.regression test fails because ../opencv_extra-master/testdata/cv/features2d/feature_detectors/detector-harris.xml.gz files has data for previous variant of goodFeaturesToTrack. Could you help me to generate new test data file for this test?
BR,
Alexander

**Comment by AlexanderUsentsov** [comment]
Hi!
I’ve update  opencv_extra and sent PR.  Unfortunately testing is not start.

Could you describe the procedure how to start testing after update testdata?

BR,
Alexander

**Comment by mshabunin** [comment]
@AlexanderUsentsov, I still can not find your PR to _opencv_extra_: [here](https://github.com/Itseez/opencv_extra/pulls). However, your branch is actually used for testing, please, see last build logs: [linux-build-3044](http://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/3044) -> [merge opencv_extra test branch](http://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/3044/steps/Merge%20opencv_extra%20with%20test%20branch/logs/stdio)

As for generating test data, as I can see in [the test](https://github.com/Itseez/opencv/blob/master/modules/features2d/test/test_detectors_regression.cpp#L211), you need to remove old validation file and run corresponding regression test - it should create the new file in this case.

**Comment by AlexanderUsentsov** [comment]
I have two branches with name good_feature : first one for  opencv, and second  for opencv_extra. When I submit code in opencv good_feature branch testing is started. You see this result for my commit on Friday.
I set new PR when I submitted new  open_extra data today but tests didn’t start.

About generating new test data for regression test.
I deleted  detector-harris.xml.gz file and ran opencv_test_features2dd.exe --gtest_filter=*Features2d_Detector_Harris.regression*
New file was created but it has size 1 KB instead of 3KB. And test cannot load data from it.
If I run opencv_test_features2dd.exe I have the same result.

Thank you,
Alexander

**Comment by mshabunin** [comment]
@AlexanderUsentsov, buildbot service does not check for changes in _opencv_extra_, thus it does not start new build when new commit arrives into this repository. You can force the build by rewriting the last commit in _opencv_ repository.
I've started new build manually.

>I have two branches with name good_feature

And you need to create two pull requests, one for _opencv_ (this one), another for _opencv_extra_.

Also, as I can see, the latest commit in your [opencv_extra/good_feature](https://github.com/AlexanderUsentsov/opencv_extra/commits/good_feature) branch was created 27 days ago. Are you sure you've pushed it today?

Can you, please, check the contents of generated file, is it xml packed in _gz_? Does it have same structure as previous?

**Comment by mshabunin** [comment]
@AlexanderUsentsov, I was able to generate test data file without any problems. It does not have _detector_params_ section because this method is not implemented in new _GFTTDetector_ class, but keypoints information is there and the test (_Harris.regression_) passes.

**Comment by AlexanderUsentsov** [comment]
I am able to generate test data for Harris.regression test now.

Thank you,
Alexander

**Comment by mshabunin** [comment]
@AlexanderUsentsov, there are some tests failures with new version:
- On MacOSX: _OCL_Imgproc/GoodFeaturesToTrack.EmptyCorners_ takes more than 180 seconds and gets killed by build service
- _OCL_GoodFeaturesToTrackFixture_GoodFeaturesToTrack.GoodFeaturesToTrack_ perf tests failed
- _OCL_PyrLKOpticalFlowFixture_PyrLKOpticalFlow.PyrLKOpticalFlow_ perf tests failed

File: modules/imgproc/src/featureselect.cpp [hunk 1]
```diff
@@ -305,15 +299,36 @@ void cv::goodFeaturesToTrack( InputArray _image, OutputArray _corners,
         for( int x = 1; x < imgsize.width - 1; x++ )
         {
             float val = eig_data[x];
-            if( val != 0 && val == tmp_data[x] && (!mask_data || mask_data[x]) )
+            if( val != 0 && (fabs(val - tmp_data[x])<CV_GFTTEPS) && (!mask_data || mask_data[x]) )
```
Comment (alalek) [inline modules/imgproc/src/featureselect.cpp]: Please revert this change. Float equality operation is fine here.

File: modules/imgproc/src/featureselect.cpp [hunk 2]
```diff
@@ -76,9 +70,9 @@ static bool ocl_goodFeaturesToTrack( InputArray _image, OutputArray _corners,
 {
     UMat eig, maxEigenValue;
     if( useHarrisDetector )
-        cornerHarris( _image, eig, blockSize, 3, harrisK );
+        cornerHarris( _image, eig, blockSize, 3, harrisK, BORDER_REPLICATE );
```
Comment (alalek) [inline modules/imgproc/src/featureselect.cpp]: Please revert border type to default here and below.

File: modules/imgproc/src/featureselect.cpp [hunk 3]
```diff
@@ -305,15 +299,36 @@ void cv::goodFeaturesToTrack( InputArray _image, OutputArray _corners,
...
+    for(i=0;i<total-1;i++)
+    {
+       for(j=i+1;j<total;j++)
+       {
+           if(*vec[j] > *vec[i])
+           {
+               pTmp = vec[i]; vec[i] = vec[j];vec[j] = pTmp;
+           }
+       }
+    }
```
Comment (alalek) [inline modules/imgproc/src/featureselect.cpp]: `std::sort` above should work fine. Revert this change.

File: modules/imgproc/src/featureselect.cpp (hunk 1 above)
Comment (AlexanderUsentsov) [inline modules/imgproc/src/featureselect.cpp]: Float equality operation is fine here but it is not correct in general. I propose  to use this approach.

File: modules/imgproc/src/featureselect.cpp (hunk 3 above)
Comment (AlexanderUsentsov) [inline modules/imgproc/src/featureselect.cpp]: Hello!

I have finished working with code for this (#3540) pull request. All test passed.
Could you review my code?

Thanks,
Alexander

File: modules/imgproc/test/test_goodfeaturetotrack.cpp [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+    test_case_count = 4;
+
+}
+
+int CV_GoodFeatureToTTest::prepare_test_case( int test_case_idx )
+{
+    const static int types[] = { CV_32FC1, CV_8UC1 };
+
+    cvtest::TS& tst = *cvtest::TS::ptr();
+    src = imread(string(tst.get_data_path()) + "stereomatching/datasets/tsukuba/im2.png", IMREAD_COLOR);
```
Comment (mshabunin) [inline modules/imgproc/test/test_goodfeaturetotrack.cpp]: I suggest to use image from "shared" folder.

File: modules/imgproc/test/test_goodfeaturetotrack.cpp [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+        TEST_MESSAGEL ("Number of feature not equal to maxCorners. RefCorners = ", Refcorners.size())
+        TEST_MESSAGE ("\n")
+    }
+    else
+    {
+        TEST_MESSAGEL ("             RefCorners = ", Refcorners.size())
+        TEST_MESSAGE ("\n")
+    }
+
+    double e =norm(corners, Refcorners);
```
Comment (mshabunin) [inline modules/imgproc/test/test_goodfeaturetotrack.cpp]: It is better to use `cvtest::norm` here.

**Comment by mshabunin** [comment]
@AlexanderUsentsov , I have two more comments:

- remove or rename the `good_feature` branch from your _opencv_extra_ fork, otherwise the CI system will not be able to merge this PR
- squash all commits into one before merge

**Comment by vpisarev** [comment]
@mshabunin, are we ready to merge it?

**Comment by mshabunin** [comment]
@vpisarev , no, not yet.

@AlexanderUsentsov , looks like the test itself is missing now. I have a backup of the previous state if you need it: [mshabunin/opencv:good_feature](https://github.com/Itseez/opencv/compare/master...mshabunin:good_feature)

**Comment by AlexanderUsentsov** [comment]
I am sorry.
Test was really missed in last commit.

BR,
Alexander

**Comment by mshabunin** [comment]
:+1;
